Modules/hmacmodule.c

```diff
@@ -0,0 +1,1713 @@
+#define Py_HMAC_HACL_UPDATE_LOOP(                                   \
+    HACL_STATE, BUF, LEN,                                           \
+    ALGORITHM, ERRACTION                                            \
+)                                                                   \
+    do {                                                            \
+        while ((Py_ssize_t)LEN > UINT32_MAX_AS_SSIZE_T) {           \
+            Py_HMAC_HACL_UPDATE_ONCE(HACL_STATE, BUF, UINT32_MAX,   \
+                                     ALGORITHM, ERRACTION);         \
+            BUF += UINT32_MAX;                                      \
+            LEN -= UINT32_MAX;                                      \
+        }                                                           \
+    } while (0)
+#else
+#define Py_HMAC_HACL_UPDATE_LOOP(   \
+    HACL_STATE, BUF, LEN,           \
+    _ALGORITHM, _ERRACTION          \
+)
+#endif
+
+/*
+ * Perform the HMAC-HASH update() operation in a streaming fashion.
+ *
+ * The formal signature of this macro is:
+ *
+ *     (HACL_HMAC_state *, uint8_t *, C integer, PyObject *, (C statements))
+ */
+#define Py_HMAC_HACL_UPDATE(                            \
+    HACL_STATE, BUF, LEN,                               \
+    ALGORITHM, ERRACTION                                \
+)                                                       \
+    do {                                                \
+        Py_HMAC_HACL_UPDATE_LOOP(HACL_STATE, BUF, LEN,  \
+                                 ALGORITHM, ERRACTION); \
+        Py_HMAC_HACL_UPDATE_ONCE(HACL_STATE, BUF, LEN,  \
+                                 ALGORITHM, ERRACTION); \
+    } while (0)
+
+/*
+ * HMAC underlying hash function static information.
+ */
+typedef struct py_hmac_hinfo {
+    /*
+     * Name of the hash function used by the HACL* HMAC module.
+     *
+     * This name may differ from the hashlib names. For instance,
+     * SHA-2/224 is named "sha2_224" instead of "sha224" as it is
+     * done by 'hashlib'.
+     */
+    const char *name;
+
+    /* hash function information */
+    HMAC_Hash_Kind kind;
+    uint32_t block_size;
+    uint32_t digest_size;
+
+    /* HACL* HMAC API */
+    py_hmac_hacl_api api;
+
+    /*
+     * Cached field storing the 'hashlib_name' field as a Python string.
+     *
+     * This field is NULL by default in the items of "py_hmac_static_hinfo"
+     * but will be populated when creating the module's state "hinfo_table".
+     */
+    PyObject *display_name;
+    const char *hashlib_name;   /* hashlib preferred name (default: name) */
+
+    Py_ssize_t refcnt;
+} py_hmac_hinfo;
+
+// --- HMAC module state ------------------------------------------------------
+
+typedef struct hmacmodule_state {
+    _Py_hashtable_t *hinfo_table;
+    PyObject *unknown_hash_error;
+    /* HMAC object type */
+    PyTypeObject *hmac_type;
+    /* interned strings */
+    PyObject *str_lower;
+
+    bool can_run_simd128;
+    bool can_run_simd256;
```

Comment:
Currently, those two are set to false explicitly. We need to use the same detection mechanism as for blake2 but I'll do it later.